src/v/archival/archival_policy.cc

```diff
@@ -10,90 +10,118 @@
 
 #include "archival/archival_policy.h"
 
+#include "archival/logger.h"
 #include "storage/disk_log_impl.h"
+#include "storage/fs_utils.h"
 #include "storage/segment.h"
 #include "storage/segment_set.h"
+#include "storage/version.h"
+#include "vlog.h"
+
+#include <seastar/core/shared_ptr.hh>
 
 namespace archival {
 
-class arch_policy_non_compacted_impl {
-public:
-    arch_policy_non_compacted_impl(
-      model::ntp ntp, const model::revision_id& rev)
-      : _ntp(std::move(ntp))
-      , _rev(rev) {}
+std::ostream& operator<<(std::ostream& s, const upload_candidate& c) {
+    s << "{ exposed_name: " << c.exposed_name
+      << ", starting_offset: " << c.starting_offset
+      << ", segment_file_name: " << c.source->reader().filename() << " }";
+    return s;
+}
 
-    std::optional<manifest> make_local_manifest(storage::log_manager& lm) {
-        manifest tmp(_ntp, _rev);
-        std::optional<storage::log> log = lm.get(_ntp);
-        if (!log) {
-            return std::nullopt;
-        }
-        auto plog = dynamic_cast<storage::disk_log_impl*>(log->get_impl());
-        if (plog == nullptr) {
-            return std::nullopt;
-        }
-        for (const auto& segment : plog->segments()) {
-            bool closed = !segment->has_appender();
-            bool compacted = segment->is_compacted_segment();
-            if (!closed || compacted) {
-                continue;
+archival_policy::archival_policy(model::ntp ntp)
+  : _ntp(std::move(ntp)) {}
+
+upload_candidate archival_policy::get_next_candidate(
```

Comment:
this function has a high axiomatic complexity (many branches). we should refactor some of the branches into functions?